### DISCUSSION TITLE
No RING when settings was changed

### DISCUSSION THREAD

**Discussion by VasylSamoilov** [IchthysMaranatha/asterisk-chan-quectel#44] [Q&A] [answered]
If quectel settings were changes, there is a possibility that module will not indicate RING and will not react to incoming calls in console.
Please consider adding `AT+QURCCFG="urcport","usbat"` to the list of initialization commands.
There are possibly some other commands that can suppress ring indication, they can be considered too.
 
Per-device user-defined init strings can also be an option, to init some quectel variants properly.

**Comment by IchthysMaranatha** [comment, marked as answer]
Generally, one time settings are not added to initialization commands, only those that are needed on every reboot.

**Reply by VasylSamoilov** [reply]
That is understandable, but proper output of ring indication is necessary for proper operation, otherwise you get undefined behavior (it may ring, or it may not). It's not one-time, it should be ensured on every reboot.

**Reply by IchthysMaranatha** [reply]
Command needed is one time, it persists on every reboot. If someone changes it, he/she should change it back to work with this again.

**Reply by VasylSamoilov** [reply]
Not the most elegant solution, but it surely can be a documentation item. Closing.

**Reply by VasylSamoilov** [reply]
Some software like quectels native Qnavigator resets this setting when initializing modem, and default value is (or maybe, depending on module) different from what is required. Implicitly setting this value during initialization can help with a lot of confusion when device is not exclusively used with chan_quectel. This setting is required to be set for proper operation.

**Comment by maisx2021** [comment]
Hello everyone,
I'm experimenting with the Chan-Quectel script on a RaspPBX running on a Raspberry Pi. It's being detected perfectly by Asterisk in UAC mode, but when I receive a call, Asterisk remains completely silent in the console. What could be the issue? Thank you very much for any assistance.
Maisx

**Reply by VasylSamoilov** [reply]
try https://github.com/RoEdAl/asterisk-chan-quectel - it's a better fork as far as I tested.

**Reply by erwan2978** [reply]
Thanks for this immediate "solution"; 
I have not yet tried these!

**Reply by erwan2978** [reply]
This day I tried the  RoEdAl fork, with quectel EC25-E, and neither incoming call, nor outgoing call was working.

I did not change anything in extensions.conf, only quectel.conf was the original provided in sources of this fork
Apparently the module from this fork set the modules in extended mode with AT+CFUN=1
Then in incoming call you get event like (asterisk stopped to be able to user ttyUSB2 by minicom);
+QIND: "act","HSDPA&HSUPA"                                                                                                                                                   
+QIND: "ccinfo",2,1,4,0,0,"+33123456789",145     (phone number changed in this example indeed) 
instead of RING in minimal mode. 
But once asterisk restarted, the program chan_quectel does not detect a call with this.

In asterisk -r I could something like
 [Apr 17 18:55:17] WARNING[328979]: src/at_response.c:1010 handle_clcc: [quectel0] CLCC (not found) idx:2 dir:1 state:6 mode:0 mpty:off number:+33123456789 type:145
                                                                                           

On the other end when trying to use outgoing rules with syntax action like 
Dial(Quectel/quectel0/${EXTENS}) , i get any channel availiable !

So I just got a bad experience with this code, and I have to go back with AT+CFUN=0
and then put back the code form git [IchthysMaranatha], and my original quectel.conf.

This is just a warning for others. Usage of this code is less obvious than the original code here

**Comment by erwan2978** [comment]
Hello
before having a try RoEdAl fork, I prefer to continue with this original development, wich does not require CMAKE
I have tested during several days the 3 main "usbnet" mode and their corresponding linux kernel module
(0: cdc_wdm/qmi_wan ; 1: cdc_ether ; 2: cdc_wdm/cdc_mbim)

I explored also the GPS data flow. And of course when you want:
VOICE call, SMS, IP DATA via APN of the GSM provider, and GPS data, like on any smart phone, you have to fight with the availability of the used ttyUSB for each task. Because therei is concurrence between Asterisk and ModemManager.
The last service require that the ttyUSB2 (when there is no other USB serial device on the PC) is reserved in something like for example 80-quectel-EC25.rules in /etc/udev/rules.d with this line :
`ATTRS{idVendor}=="2c7c", ATTRS{idProduct}=="0125", ENV{.MM_USBIFNUM}=="02", ENV{ID_MM_DEVICE_IGNORE}="1"
`
(the USBIFNUM is not the n of global result /dev/ttyUSBn, but one of the 0, 1, 2 ,3 pseudo serial associated to quectel modem (EC25-E in my experience)

In all this experiments between the 3 "usbnet" config listed above, at a moment the voice output flow through tty (gps) was lost. I circumvent the problem with uac mode.
 = = = =
But over all , I could notice the problem discussed here : some time the incoming call was not detected.
Very soon each time I got this problem, I stopped asterisk, and launched minicom, on the ttyUSB used by asterisk to manage AT command. So the first tiem I tested asterisk, was when I was using cdc_ether ("usbnet",1)
The incoming call was signaled by event:
` +CRING: VOICE`

so I changed that in at_response.h in the definition of AT_RESPONSES_TABLE(_):
`        _( RING,        "RING",         "RING\r") \`
in
`        _( RING,        "RING",         "+CRING: VOICE\r") \`

= = = =
A possible approach  would be something like the solution for the command "accept the incoming call"
with a pointer on string list:
like at_command.h, in the definition of  AT_COMMANDS_TABLE(_); on line:
`        _( AT_A,            "at+qpcmv=0;+qpcmv=1,0;a")`

So the goal is to detect the event for both possible value in the function at_response;   in code at_response.c:
```
                        case RES_RING:
                                return at_response_ring (pvt);
```

But with brief look, I do not see how to do it without breaking all the code. I'm not a developer, so it takes to me a lot of time to understand how to manipulate pointers
If no one sees how to correct easily this, may be I shall try a bad hack: duplicate the RING entry in AT_RESPONSE_TABLE; a second line _(RING2 ....., and a duplicate case RES_RING2 in at_response.c ???

**Reply by IchthysMaranatha** [reply]
Do you get both types (RING and CRING) for the same call? If not, create both response cases as you've said and point to same function in at_response.c

**Reply by erwan2978** [reply]
Fortunately not in the same CALL.
I just noticed one or the other after a change in the mode , or in some parameters. I do not change any parameters, the message during CALL remain the same. So indeed when I shall not have any change to do, a last compilation with the current format, RING or "+CRING: VOICE", in at_response.h is a convenient solution.
I shall try tomorrow, and make a notice here, once the test are done.

**Reply by erwan2978** [reply]
OK so I made the test with that both patches:

```
--- ../asterisk-chan-quectel.IchthysMaranatha/at_response.h     2025-04-08 17:57:26.835446118 +0200
+++ at_response.h       2025-04-08 10:28:04.708377614 +0200
@@ -50,6 +50,7 @@
        _( OK,          "OK",           "OK\r") \
        _( CONF,        "^CONF",        "MISSED_CALL:") \
        _( RING,        "RING",         "RING\r") \
+       _( RING2,       "RING",         "+CRING: VOICE\r") \
 \
        _( RSSI,        "^RSSI",        "^RSSI:") \
        _( SMMEMFULL,   "^SMMEMFULL",   "^SMMEMFULL:") \
```

```
--- ../asterisk-chan-quectel.IchthysMaranatha/at_response.c     2025-04-08 17:57:26.835446118 +0200
+++ at_response.c       2025-04-08 10:28:04.708495464 +0200
@@ -2135,6 +2135,9 @@
                        case RES_RING:
                                return at_response_ring (pvt);
 
+                       case RES_RING2:
+                               return at_response_ring (pvt);
+
                        case RES_SMMEMFULL:
                                return at_response_smmemfull (pvt);
 /*
```

I tested the two pattern maching in at_response.h, for RING:
"+CRING: VOICE\r"
and
"+CRING:"

With "CRING:" the call is not detected by chan_quectel.so.

Both are working; chan_quectel.so detects RING event
I keep the first because I have no idea why there this extra qualifier VOICE ! Is this because we could have something like FAX ? 
I have no idea about what kind of service are offered by some Mobile Operator in the world, so if here in France according to my knowledge, no Mobile provider offer FAX on GSM ?
But for sure "+CRING:", without any extra qualifier, is more common. 

On Huawei ME909s-821, I could see also, either RING, or +CRING: during incoming call test with ttyUSB2 monitoring with minicom;
 without having search how to get this indication format. I just see that but without having done something deterministic to get such a behavior.
So now could you include this in your code ? Not every body knows or tries to debug the code. So some just notice here that the incoming call are not detected, and then conclude that the code is NOT for their hardware, and in some case search a solution elsewhere.
For my case with just all the last little change you accepted these last days, all is working for my whished target:
Voice call, with "good" quality with uac mode, exchange SMS, IP data with cdc_ether ( AT+QCFG="usbnet",1  ), 

And also at the same time than all the service above, GPS data flow on /dev/ttyUSB1, when you do not use it anymore for voice purpose because you are using alsa driver with uac !
On /dev/ttyUSB2 while asterisk is stopped temporarily
AT+QGPSCFG="outport","usbnmea"
AT+QGPSCFG="gpsnmeatype",3
AT+QGPSGNMEA="RMC"       (for exemple , see Quectel_EC25EC21_GNSS_AT_Commands_Manual_V1.1.pdf for other output format)

open minicom on /dev/ttyUSB1, or gpsd on this device and look with GUI xgps
AT+QGPS=1 (to activate the GPS data flow)
the flow start each second

later use
AT+QGPSEND
 to close the data flow
====
NOTICE: in this GPS data flow, on ttyUSB1 I see "NO CARRIER" event each time you close communication
Thanks for this convenient fork !

**Reply by erwan2978** [reply]
just a extra notice for readers here
Of course without stopping asterisk, or ModemManager, you can ask to asterisk to use its locked ttyUSB2, to perform a single AT command:

root@:~#: **asterisk -x 'quectel cmd quectel0 AT+QGPS=1'**
[quectel0] 'AT+QGPS=1' AT command queued
Asterisk ending (0)

root@:~# **asterisk -x 'quectel cmd quectel0 AT+QGPSEND'**
[quectel0] 'AT+QGPSEND' AT command queued
Asterisk ending (0).

**Reply by IchthysMaranatha** [reply]
Try this
in at_response.h
` _( RING2,       "+CRING",         "+CRING:") \`
in at_response.c call new function for RING2 
~~~
     case RES_RING2:
                 return at_response_cring (pvt, str);
~~~
create new function at_response_cring below at_response_ring
~~~
static int at_response_cring (struct pvt* pvt, const char* str)
{
 char cr[10];
  if (sscanf (str, "+CRING: VOI%s", cr) == 1)
    {
	if (pvt->initialized)
	{
		pvt->ring = 1;
		pvt->dialing = 0;
		pvt->cwaiting = 0;

		pvt->rings++;

		request_clcc(pvt);

		/* We only want to syncronize volume on the first ring and if no channels yes */
/*		if (pvt->volume_sync_step == VOLUME_SYNC_BEGIN && PVT_NO_CHANS(pvt))
		{
			if (at_enqueue_volsync(&pvt->sys_chan))
			{
				ast_log (LOG_ERROR, "[%s] Error synchronize audio level\n", PVT_ID(pvt));
			}
			else
				pvt->volume_sync_step++;
		}
*/
	}
    }
	return 0;
}
~~~

**Reply by IchthysMaranatha** [reply]
Oh sorry, you've already got it working, right? Good, well done!

**Reply by IchthysMaranatha** [reply]
The loss of RING indicator is due to some other software (like modem manager) changing the settings of the modem. For eg if you run these commands after you lose the RING urc
~~~
AT+QINDCFG="all", 1
AT+QINDCFG="ring", 1
~~~
you should get it back. Which is why I'm a little hesitant to incorporate changes which adjusts for the module not being used as a dedicated voice module for asterisk as there might be no end to it. But I'm tempted to add your cring code as it's only a couple of lines.

**Reply by erwan2978** [reply]
Thank for your reply.
If you do not change the code, an other solution, is to advertise users, one the home directory of the code, among others recommendation;
like
_If you have no sound in outgoing way, thinks about this solution ..._

I just recall to you that my first goal  when I bought a GSM modem, was to get a recovery route, when my main FTTH line is down. An only after months using it like that, I just think about sending to me on my main mobile, SMS, when FTTH line is down. And after I thought :   but why not using it too for VOICE, with my already existing asterisk!
Notice also, that not using ModemManager, means usage of cdc_ether is mandatory (usbnet",1 , profile of EC25)
Personally on Fedora I do not know, how to use the both other modes (cdc_wdm with cdc_mbim, or qmi_wan) without the help of ModemManager. Maybe there are solution on Debian (with /etc/networks/interfaces ?), but not on Fedora or RedHat like linux

**Reply by IchthysMaranatha** [reply]
Using cdc_ether is not the issue here, I think modem manager or something else that you use sets AT+CRC=1, which leads to the extended ring indication with CRING. Instead of using commands I wrote earlier, using AT+CRC=0 will give back the RING indication in this case. If you just prevent that at+crc command, you won't have this issue.

**Reply by erwan2978** [reply]
I just  made  a reboot of my VM router+asterisk this morning, because  during night the FTTH line was not, available, and until now not yet available:
And after checking that asterisk was working right, I stopped the service and made test on /dev/ttyUSB2 with minicom
and without having done nothing, I got this:
                                                                                              
OK                                                                                                     
**AT+QINDCFG="all"**                                                                                       
+QINDCFG: "all",1                                                                                      
                                                                                                       
OK                                                                                                     
**AT+QINDCFG="ring"**                                                                                      
+QINDCFG: "ring",1                                                                                     
                                                                                                       
OK 
**AT+CRC?**                                                                                                
+CRC: 0                                                                                                
                                                                                                       
OK
And of course, when I perform call, I get 
RING
on ttyUSB2
So all seems  right accordingly to your suggestion.
But yesterday, I was gettng
+CRING: VOICE

So now I let in the code using by me, modification I suggested to you in at_response.h and at_responce.c, because I want a reliable asterisk behaviour. The problem is not my case, but the practical experimentation that many users will have with the code you let in line.
Unfortunately; not every body  give bak information about their experience, so its difficult to make true statistics.
Probably some made silently modification to get a running code without trouble. Others pass their way with this code.

**Reply by IchthysMaranatha** [reply]
Just try and investigate what caused AT+CRC=1 on your modem, only when that command is passed urc changes to CRING. 

All users facing even the same issue do not have the same cause. So your solution would not work for someone whose RING indicator has gone off for a different reason, he may not even get CRING. You saw another user in another thread had a problem because he had not checked with aplay what his alsa device was actually named and used that in the conf. But he had a one way audio issue which you also faced earlier due to some other cause. The very first post in this thread also refers to changed settings when using some other software. Catering to all of these has no end. Users have to have a certain degree of understanding and competence when using OSS, everything cannot be troubleshot in code adjusting for every possible scenario.

**Reply by erwan2978** [reply]
Yes, this is a valuable point of view.
The most important here, is that we have this discussion. I found this thread with google, in the five first suggestions when you search no ring response in asterisk. So people will arrive here, and will have some way to investigate.
Thanks for your help

**Reply by erwan2978** [reply]
Finally, I definitively stopped and disabled ModemManager service on my Fedora Linux System. 
This means, no other solution than **cdc_ether** is usable ,for IP Data over 4G, for Fedora System.
So now I let the  quectel Module in this configuration (AT+QCFG="usbnet",1 ), plus the configuration you suggested above.
Then since more than a month now, I never get a single bug either on my IP Data mobile interface; neither on my asterisk behavior. And now I get only the "RING\n" only indicator, on incoming call event
All this: IP Data on GSM, Asterisk with UAC mode for voice data, SMS, are working fine since then, 24h/24 7d/7.
A real professional system !
